dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -3048,78 +3048,131 @@ RtpsUdpDataLink::RtpsWriter::send_and_gather_nack_replies(MetaSubmessageVec& met
+            const RtpsUdpSendStrategy::OverrideToken ot =
+              link->send_strategy()->override_destinations(addrs);
+            send_buff_->resend_i(SequenceRange(seq, seq), 0, reader->id_);
+            continue;
+          }
         }
-        if (Transport_debug_level > 5) {
-          ACE_DEBUG((LM_DEBUG, "(%P|%t) RtpsUdpDataLink::RtpsWriter::send_and_gather_nack_replies "
-                     "local %C remote %C requested resend\n",
-                     LogGuid(id_).c_str(),
-                     LogGuid(info->id_).c_str()));
+
+        if (reader->durable_) {
+          OPENDDS_MAP(SequenceNumber, TransportQueueElement*)::const_iterator diter =
+            reader->durable_data_.find(seq);
+          if (diter != reader->durable_data_.end()) {
+            // Durable.
+            link->durability_resend(diter->second);
+            continue;
+          }
+
+          if (reader->durable_timestamp_.is_zero()) {
+            // Ignore the request, i.e., don't send a gap because we don't have the durable data yet.
+            continue;
+          }
+        }
+
+        if (durable_) {
+          // Must send directed gap.
+          gaps.insert(seq);
+        } else {
+          // Non-directed gap.
+          consolidated_recipients.insert(addrs.begin(), addrs.end());
+          consolidated_gaps.insert(seq);
+          continue;
         }
       }
-      info->requested_changes_.clear();
     }
+
+    // Gather directed gaps.
+    gather_gaps_i(reader->id_, gaps, meta_submessages);
   }
   readers_expecting_data_.clear();
 
-  DisjointSequence gaps;
-  if (!requests.empty()) {
-    if (send_buff_.is_nil() || send_buff_->empty()) {
-      if (Transport_debug_level > 5) {
-        ACE_DEBUG((LM_DEBUG, "(%P|%t) RtpsUdpDataLink::RtpsWriter::send_and_gather_nack_replies: "
-                   "gaps = requests\n"));
-      }
-      gaps = requests;
-    } else {
-      OPENDDS_VECTOR(SequenceRange) ranges = requests.present_sequence_ranges();
-      SingleSendBuffer& sb = *send_buff_;
-      ACE_GUARD(TransportSendBuffer::LockType, guard, sb.strategy_lock());
-      const RtpsUdpSendStrategy::OverrideToken ot =
-        link->send_strategy()->override_destinations(recipients);
-      for (size_t i = 0; i < ranges.size(); ++i) {
-        if (Transport_debug_level > 5) {
-          ACE_DEBUG((LM_DEBUG, "(%P|%t) RtpsUdpDataLink::RtpsWriter::send_and_gather_nack_replies: "
-                     "resend data %q-%q\n", ranges[i].first.getValue(),
-                     ranges[i].second.getValue()));
-        }
-        sb.resend_i(ranges[i], &gaps);
-      }
+  // Send the consolidated requests.
+  const OPENDDS_VECTOR(SequenceRange) ranges = consolidated_requests.present_sequence_ranges();
+  ACE_GUARD(TransportSendBuffer::LockType, guard, send_buff_->strategy_lock());
+  const RtpsUdpSendStrategy::OverrideToken ot =
+    link->send_strategy()->override_destinations(consolidated_recipients);
```

Comment:
Can the scope of both of these objects be reduced to only over the `resend_i` phase?

Reply:
I think I resolved this.  There is still one loop where the lock and token are outside of the loop.  I'm not sure there is an advantage moving them inside the loop.